### ISSUE TITLE
freebsd support

### ISSUE THREAD

**Issue by dch** [ityonemo/zigler#69]
I think you can have FreeBSD support already:

I have the following packages installed:

```
# pkg install -r FreeBSD lang/zig lang/erlang-runtime22 devel/elixir-hex devel/llvm80
```

This gives us zig-0.5.0 and the rest should be self explanatory.

I suppressed the side-effect-only Logger https://github.com/ityonemo/zigler/blob/master/lib/zigler/compiler.ex#L24

```
$ export NIFTY_LLVM_VERSION=80
$ mix do deps.get --all, deps.compile, compile
$ mix zigler.get_zig 0.5.0
...
$ mix test --trace

Compiling 17 files (.ex)

13:19:58.168 [info]  compiling using command: `/repos/zigler/zig/zig-freebsd-x86_64-0.5.0/zig build-lib zig_nif.zig -dynamic --disable-gen-h --override-lib-dir /repos/zigler/zig/zig-freebsd-x86_64-0.5.0/lib/zig`

13:19:58.176 [info]  compiling using command: `/repos/zigler/zig/zig-freebsd-x86_64-0.5.0/zig build-lib zig_nif.zig -dynamic --disable-gen-h --override-lib-dir /repos/zigler/zig/zig-freebsd-x86_64-0.5.0/lib/zig`

13:20:04.925 [info]  compiling using command: `/repos/zigler/zig/zig-freebsd-x86_64-0.5.0/zig build-lib zig_nif.zig -dynamic --disable-gen-h --override-lib-dir /repos/zigler/zig/zig-freebsd-x86_64-0.5.0/lib/zig --release-safe`
Generated zigler app

ZiglerTest.CompilerErrorParserTest
  * test the compiler line parser (0.04ms)

ZiglerTest.CompilerErrorTest
  * test compiler error catches and rewrites internal error (270.6ms)
  * test compiler error catches mismatched arity (35.3ms)
  * test compiler error catches and rewrites external error (485.7ms)
  * test compiler error catches mismatched nif (7.9ms)

ZiglerTest.ParserTest
  * test the nif parser can identify nifs (0.01ms)
  * test the docstring multiline parser can identify multiline docstrings (0.05ms)
  * test the function header parser can identify bool function headers (0.07ms)
  * test the function header parser can identify one-arity function headers (0.04ms)
  * test the docstring line parser will ignore a non docline (0.00ms)
  * test the function header parser can identify zero-arity function headers (0.02ms)
  * test the docstring line parser can identify a single docline (0.02ms)
  * test the function header parser can identify two-arity function headers (0.05ms)
  * test the function header parser can identify beam.env headers (0.05ms)
  * test the docstring line parser can identify a single docline when it's got indentation (0.01ms)
  * test the function header parser can identify string function headers (0.04ms)
  * test the function header parser can identify erlnifenv headers (0.06ms)

ZiglerTest.UnitParserTest
  * test the identifier parser fails on non_identifiers (0.00ms)
  * test the test header parser can identify test headers (0.08ms)
  * test the identifier parser can correctly identify identifiers (0.01ms)
  * test the import parser leaves pub imports alone (0.02ms)
  * test the identifier parser fails on strange symbols (0.00ms)
  * test the import parser can identify import statements (0.01ms)

ZiglerTest.CodeTest
  * test when given a code multiline string the code module can return plain code with no nifs, and assign files and lines correctly (0.04ms)
  * test when given a code multiline string the code module can do the correct thing with multiple functions (0.1ms)
  * test when given a code multiline string the code module can identify more complex nif headers (0.09ms)
  * test when given a code multiline string the code module can find a nif and correctly assign it, adding line numbers (0.07ms)

ZiglerTest.Types.EmptyTest
  * test zero arity zig functions work (0.06ms)

ZiglerTest.MutationTest
  * test passing binaries across a function self-call works (0.07ms)
  * test we can mutate binaries (0.00ms)

ZigTest.TypespecTest
  * test zigler correctly creates typespecs for degenerate (2.0ms)
  * test zigler correctly creates typespecs for generic beam types (23.0ms)
  * test zigler correctly creates typespecs for floats (0.5ms)
  * test zigler correctly creates typespecs for integers (0.9ms)
  * test zigler correctly creates typespecs for bools (0.7ms)
  * test zigler correctly creates typespecs for float lists (0.4ms)
  * test zigler correctly creates typespecs for strings (0.7ms)
  * test zigler correctly creates typespecs for int lists (2.8ms)

ZiglerTest.ZigdocTest
  * test gets the docs (695.7ms)

ZiglerTest.AllocatorsTest
  * test elixir persistent memory works (0.1ms)
  * test elixir basic allocator works (0.1ms)

ZiglerTest.DoubleZigTest
  * test two zig definitions in the different sigils (0.06ms)
  * test two zig definitions in the same sigil (0.05ms)

ZiglerTest.ExternalFilesTest
  * test a recursive external file can be called (0.05ms)
  * test a relative external file can be called (0.04ms)

ZiglerTest.Types.EgressTest
  * test lists can be egressed from c_int lists (0.1ms)
  * test lists can be egressed from internal literals (0.00ms)
  * test integers can be egressed c_int (0.00ms)
  * test lists can be egressed from f16 lists (0.00ms)
  * test lists can be egressed from i32 lists (0.00ms)
  * test floats can be egressed f32 (0.00ms)
  * test floats can be egressed f64 (0.00ms)
  * test lists can be egressed from c_long lists (0.00ms)
  * test strings can be egressed from a slice (0.06ms)
  * test integers can be egressed i32 (0.00ms)
  * test integers can be egressed c_long (0.00ms)
  * test lists can be egressed from i64 lists (0.00ms)
  * test bools can be exported correctly (0.00ms)
  * test characters can be egressed correctly (0.00ms)
  * test strings can be egressed from a c_string (0.00ms)
  * test lists can be egressed from f32 lists (0.00ms)
  * test lists can be egressed from f64 lists (0.00ms)
  * test integers can be egressed i64 (0.00ms)

ZiglerTest.Types.IngressTest
  * test 32-bit integers can be ingressed and are guarded for invalid values (3.5ms)
  * test binaries can be ingressed as slices and are guarded for invalid values (0.09ms)
  * test atoms can be ingressed and are guarded for invalid values (0.02ms)
  * test f32 lists can be ingressed correctly (0.07ms)
  * test pids can be ingressed as beam.pid correctly (0.00ms)
  * test f16 lists can be ingressed and are guarded for invalid values (0.09ms)
  * test binaries can be ingressed as binary structs and are guarded for invalid values (0.01ms)
  * test pids can be ingressed as e.ErlNifPid correctly (0.00ms)
  * test binaries can be ingressed as binary structs via e.ErlNifBinary and are guarded for invalid values (0.01ms)
  * test i32 lists can be ingressed correctly (0.00ms)
  * test 16-bit floats can be ingressed and are guarded for invalid values (0.01ms)
  * test f64 lists can be ingressed even with empties (0.00ms)
  * test binaries can be ingressed as slices correctly (0.00ms)
  * test atoms can be ingressed correctly (0.00ms)
  * test 16-bit floats can be ingressed correctly (0.00ms)
  * test pids can be ingressed as beam.pid and are guarded for invalid values (0.01ms)
  * test f32 lists can be ingressed and are guarded for invalid values (0.02ms)
  * test i64 lists can be ingressed even with empties (0.00ms)
  * test i64 lists can be ingressed and are guarded for invalid values (0.02ms)
  * test binaries can be ingressed as c_strings correctly (0.00ms)
  * test usize can be ingressed correctly (0.00ms)
  * test 64-bit integers can be ingressed correctly (0.00ms)
  * test binaries can be ingressed as binary structs correctly (0.00ms)
  * test c integers can be ingressed and are guarded for invalid values (0.01ms)
  * test 32-bit integers can be ingressed correctly (0.00ms)
  * test c integers can be ingressed correctly (0.00ms)
  * test f32 lists can be ingressed even with empties (0.00ms)
  * test f16 lists can be ingressed correctly (0.00ms)
  * test binaries can be ingressed as binary structs via e.ErlNifBinary correctly (0.00ms)
  * test i32 lists can be ingressed and are guarded for invalid values (0.02ms)
  * test i32 lists can be ingressed even with empties (0.00ms)
  * test usize can be ingressed and are guarded for invalid values (0.02ms)
  * test c longs can be ingressed and are guarded for invalid values (0.02ms)
  * test 64-bit floats can be ingressed correctly (0.00ms)
  * test 64-bit integers can be ingressed and are guarded for invalid values (0.02ms)
  * test f64 lists can be ingressed correctly (0.00ms)
  * test bools can be ingressed and are guarded for invalid values (0.02ms)
  * test characters can be ingressed correctly (0.00ms)
  * test pids can be ingressed as e.ErlNifPid and are guarded for invalid values (0.01ms)
  * test f16 lists can be ingressed even with empties (0.00ms)
  * test 32-bit floats can be ingressed correctly (0.00ms)
  * test isize can be ingressed and are guarded for invalid values (0.01ms)
  * test c longs can be ingressed correctly (0.00ms)
  * test binaries can be ingressed as c_strings and are guarded for invalid values (0.01ms)
  * test f64 lists can be ingressed and are guarded for invalid values (0.02ms)
  * test 64-bit floats can be ingressed and are guarded for invalid values (0.02ms)
  * test 32-bit floats can be ingressed and are guarded for invalid values (0.01ms)
  * test characters can be ingressed and are guarded for invalid values (0.02ms)
  * test bools can be ingressed correctly (0.00ms)
  * test isize can be ingressed correctly (0.00ms)
  * test i64 lists can be ingressed correctly (0.00ms)

ZiglerTest.ReadmeTest
  * test allocations (0.1ms)
  * test example_zig (0.05ms)
  * test zig_collections (0.09ms)

ZiglerTest
  * test sigil-Z2 creates a function of the correct arity (0.05ms)
  * test sigil-Z creates a public function (0.06ms)

ZiglerTest.ZigTest
  * zigtest the truth (0.00ms)

ZiglerTest.ZigFailingTest
  * test if we have a failing test, it can be caught (5.8ms)

ZiglerTest.ZigIndirectTest
  * zigtest function one (0.00ms)

ZiglerTest.ZigDeepTest
  * zigtest function one (0.00ms)


Finished in 31.5 seconds
120 tests, 3 zigtests, 0 failures

Randomized with seed 960707
```

Which looks *perfect* to me. It would be nice if zigler knew to use the pre-installed `lang/zig` port, rather than needing to download stuff.

**Comment by ityonemo** [comment]
@dch 

Amazing.  It's just so great when software just works.  I'm going to push an 0.1.1 with freebsd support.  

1) I'm afraid that it's been about 10 years since I have used FreeBSD... does "which" work?  Because if it does I'm happy to just use "which" to determine the path of the local zig version, and that ought to work for linux too.

2) Is there a public CI system for freebsd?  I think elixir now uses Cirrus, which supports it, would be happy to get an account for this project.

**Comment by feld** [comment]
FreeBSD has "which".

Unsure on public CI

**Comment by ityonemo** [comment]
reviewed the option of using a "local" version of zig.  The executable itself is fine, but I'm unsure of where zig will put its standard library code.  In the tar.gz it lives in lib/.  Dismayingly, in linux, this might not be in a consistent path.  `usr/lib/`? `usr/local/lib`?, and it may depend on what packaging system you use.  Will explore further in: https://github.com/ityonemo/zigler/issues/72

implemented freebsd enablement in 7a8fc56252f6e48fcb6f42729016c666edaaa1b
